### ISSUE TITLE
Bindless Tracking Issue

### ISSUE THREAD

**Issue by kanerogers** [gfx-rs/wgpu#3637]
# Overview
This issue tracks enabling various "bindless" functionality for the various native backends.

For a high level guide on what we believe the bindless api should look like, look at https://hackmd.io/@cwfitzgerald/wgpu-bindless

## Binding Array Support
- [x] https://github.com/gfx-rs/wgpu/pull/711
- [x] https://github.com/gfx-rs/wgpu/pull/715
- [x] https://github.com/gfx-rs/wgpu/issues/1635
- [x] https://github.com/gfx-rs/wgpu/issues/3334
- [ ] #6741
- [ ] #6744
- [ ] #6739
- [x] #6743
- [x] #6737
- [x] https://github.com/gfx-rs/wgpu/issues/3249

#### Uniform Buffers
- [ ] https://github.com/gfx-rs/wgpu/issues/7149
- [ ] #6740
- [ ] #6742 
- [ ] #6733

#### Partially bound descriptors
- [x] https://github.com/gfx-rs/wgpu/issues/1635
- [x] #3639 
- [x] #3640 

## Validation
- [ ] https://github.com/gfx-rs/wgpu/issues/3648
- [ ] https://github.com/gfx-rs/wgpu/issues/3692
- [ ] https://github.com/gfx-rs/wgpu/issues/5102
- [x] #6736
- [x] #6738

## GPU Validation
- [ ] Validate Binding Array Access is in Bounds on Vulkan
- [ ] Validate Binding Array Access is in Bounds on Metal
- [ ] Validate Binding Array Access is in Bounds on DX12

## [Sparse Bind Groups](https://hackmd.io/@cwfitzgerald/wgpu-bindless?stext=2938%3A1305%3A0%3A1733977963%3AQCxWFF)
- [ ] Implement Validation Resource Metadata Buffer on Vulkan
- [ ] Implement Validation Resource Metadata Buffer on Metal
- [ ] Implement Validation Resource Metadata Buffer on DX12

## [Mutable Bind Groups](https://hackmd.io/@cwfitzgerald/wgpu-bindless#Updating-Bind-Groups)
- [ ] Implement `BindGroup::update_bindings` Without Holes

## [Read Only Resources](https://hackmd.io/@cwfitzgerald/wgpu-bindless#Read-Only-Resources)
- [ ] Implement `Texture::set_usages`
- [ ] Implement `Buffer::set_usages`

## [Temporary Removal](https://hackmd.io/@cwfitzgerald/wgpu-bindless#Temporary-Removal)

- [ ] Implement prototype temporary remove api. Depends on "Sparse Bind Groups"

## Driver Bugs(?)
- [ ] https://github.com/gfx-rs/wgpu/issues/6745

**Comment by John-Nagle** [comment]
Bindless mode means you have a big table of all the texture objects in the GPU. Each of those texture objects is a GPU buffer. Buffers either belong to the GPU, and the CPU can't touch them, or they belong to the CPU, and the GPU can't touch them.

So the big table of texture object descriptors and the GPU buffer states have to be kept in sync. At the Vulkan level, this problem belongs to the program calling Vulkan. [Vulkano, which is supposed to be a safe Rust interface to Vulkan, has machinery for maintaining that table.](https://github.com/vulkano-rs/vulkano/blob/9e21fc1544290cae7e37824b2594f9020a3078a2/vulkano/src/descriptor_set/mod.rs#L309) But WGPU doesn't have that machinery, so it can't do bindless yet, at least not safely.

GPU/CPU buffer ownership and the bindless descriptor table need to be managed together.

All the targets WGPU currently supports seem to support bindless mode. Even OpenGL, since 2013, has offered it as an extension. You do need OpenGL 4 with extensions, though. Everything in current release seems to have full support now, via Vulkan, Metal, or OpenGL.

The future is bindless. Unreal Engine is now bindless-only, I think.

**Comment by SK83RJOSH** [comment]
> Unreal Engine is now bindless-only, I think.

It is not, in fact. Bindless resources are largely experimental and only enabled by default under Vulkan. They are optionally enabled under DX12 when using SM6 + ray tracing. There's still ongoing work to add support for bindless resources to shader graph as well.

Aside from that, it seems very much like there are targets that they're interested in that do not support bindless. So many materials will likely still need to implement both paths.

**Comment by cwfitzgerald** [comment]
I have filled out the above issue with the current plan for bindless, and what work has been done previously

**Comment by John-Nagle** [comment]
Right. I see more of the problems now. Looking at [Bindless Investigation and Proposal](https://hackmd.io/PCwnjLyVSqmLfTRSqH0viA), it's clear that driver-controlled residency and big arrays of bindless descriptors do not play well together. This is a non-problem for Vulkan, where all assets must be resident in GPU memory. For Metal, there is a [residency control API.](https://developer.apple.com/documentation/metal/mtlresidencyset). Not sure what the plan is for WebGPU, since that's still being defined.

I look at this from the viewpoint of needing game-type performance on large scenes, for target machines comparable to what the average Steam user has. In my own applications, I'm managing residency at the application level, where I switch textures to lower resolutions when memory is tight. Rejection of a buffer allocation request is a normal event which results in LOD reductions. For the Vulkan case, this substitutes for driver initiated eviction.

I've mostly looked at the Vulkan case, where VK_DESCRIPTOR_BINDING_UPDATE_AFTER_BIND_BIT is available and one big array of descriptors is possible. When everything is resident, that leads to simple implementations which don't require barriers during drawing, while allowing concurrent updating. I've written some design materials on that. The basic concept is that the wrapper level (WGPU) takes ownership of a buffer when it is in the descriptor table, which protects it from being deleted while rendering is in progress.

The cram job required to run on smaller targets makes things far more complicated. Driver-initiated eviction really complicates things. Can that be handled without crippling performance on the more powerful targets?

**Comment by cwfitzgerald** [comment]
To clarify, none of these apis deal with residency at all . The assumption is that all resources that are bound to bind group are resident always. Yes we need to prevent metal from making resources non-resident, but with residency sets this should be easy.

> When everything is resident, that leads to simple implementations which don't require barriers during drawing, while allowing concurrent updating

Barrier generation has basically nothing to do with residency. The purpose of the `buffer.set_usages` API, is to allow wgpu to not need to check if every single resource needs a barrier. Those resources with only read only usages can be ignored and just bound indirectly as part of the binding array

**Comment by John-Nagle** [comment]
OK, if everything is resident, things are simpler.

So, as you see bindless, what has to be done on a per-draw basis? That's where the overhead comes from. Is there checking that has to be done at each draw, or can the checking be hoisted to once per frame, or once per texture change, or to compile time?

**Comment by John-Nagle** [comment]
If an allocator is needed for descriptor slots, feel free to use [this one I wrote.](https://github.com/John-Nagle/rust-vulkan-bindless/blob/main/alloc/src/bitalloc.rs) It's lock-free.

**Comment by cwfitzgerald** [comment]
> If an allocator is needed for descriptor slots, feel free to use [this one I wrote.](https://github.com/John-Nagle/rust-vulkan-bindless/blob/main/alloc/src/bitalloc.rs) It's lock-free.

This would be deligated to the user, but good to have in case a user needs it

**Comment by John-Nagle** [comment]
Ah. You may not want to delegate that to the user. WGPU already has a buffer allocator, and descriptor slots and buffers need to be closely coordinated. I've been looking at designs which work something like this:

- Application/renderer layer requests a descriptor slot. They get back an opaque handle.
- Application/renderer layer requests a texture buffer with data. Buffer is created and filled.
- Application/renderer layer attaches the buffer to the descriptor slot handle. Descriptor manager takes ownership of the buffer, preventing it from being changed or deleted by the application. This queues insertion into the descriptor array.
- Dropping the  descriptor slot handle queues deletion from the descriptor array. 
- All descriptor array updates are processed at a point in the render cycle where the GPU is not rendering.

The idea is to use Rust ownership to manage most of the interlocking. If you let the application mess with the descriptor array, you need more checking in the lower layers.  More machinery to implement.

Bindless mode exists to improve performance by spending less time doing binding. It's only useful if it provides a big reduction in binding overhead.

Comments?

**Comment by magcius** [comment]
> So, as you see bindless, what has to be done on a per-draw basis? That's where the overhead comes from. Is there checking that has to be done at each draw, or can the checking be hoisted to once per frame, or once per texture change, or to compile time?

Nothing needs to be done on a per-draw basis (unless you count the usage of our indirection buffers). We designed it in a way that if the user uses the API properly (that is, marking resources as read-only when they are going to basically be read-only by the application), that there's very little cost to the validation. Assuming we don't find a flaw in our plan, that is :)

For read-write bindless resources, there's some validation required that needs to be done at set_bind_group time, and some barriers might need to get emitted at submit time, but we expect that the number of read-write resources to be relatively small.

> Ah. You may not want to delegate that to the user.

However, we discussed several ideas in the WebGPU WG, and currently believe that leaving bind group indices to the user hits a nice balance between user flexibility and performance.

> The idea is to use Rust ownership to manage most of the interlocking. If you let the application mess with the descriptor array, you need more checking in the lower layers. More machinery to implement.

Your library sounds like a great utility library that we might want to recommend to users of wgpu in Rust! 

Note that the bindless proposal we have provided above is perfectly safe and requires no synchronization with the GPU; shadow copies are made under the hood when updating bind groups.

That said, if we think the cost of mutating bind groups through the shadow copy remains too high, we might go the other way and handle slot allocation for the user. Whatever happens, we want to make sure that wgpu remains aligned with the WebGPU WG and specification.

**Comment by John-Nagle** [comment]
> if the user uses the API properly

Where is the new bindless API documented? Even if it's not working, I'd like to see how the API is supposed to be used.

**Comment by magcius** [comment]
The two proposals (https://hackmd.io/PCwnjLyVSqmLfTRSqH0viA and https://hackmd.io/@cwfitzgerald/wgpu-bindless) should paint a somewhat complete picture of the new API, though you might have to read between the lines a bit. It's meant for implementers.

**Comment by John-Nagle** [comment]
_Bind groups are updated on the CPU timeline and updates "immediately". ... What** this means that all previous uses of the bind group continue to use the old contents and any new usages use the updated contents._

So the descriptor table is double-buffered? Reasonable. I was thinking in terms of an update queue applied at end of frame, but that's functionally equivalent. 

"The CPU timeline" concept needs to be clarified for multi-thread programs. There are potential locking bottlenecks. As noted, this is weird for multi-thread programs.

_Note that this means that every update_bindings call will require us to make a shadow copy of all descriptors in the bind group, and associated tracking data. While we don't expect update_bindings to require a lot of memory compared to buffers and textures, it is still not a cheap operation._

Right. That's why I was thinking in terms of an update queue. Number of changes per frame is probably < 100.  Number of descriptors is on the order of 100,000 for a complex scene. But either way will work.

_By allowing a resource to be shifted to a read-only state, we let the tracking systems only worry about the resource being alive, not their state. This allows bindless arrays to be bound with very low costs._

Right. Most content is read-only. Read-write content is mostly rendering intermediates. 

What prevents freeing a bound buffer? Something has to interlock against that. Will that be an error or an operation that is deferred until it is safe?

Is a shader accessing an unused descriptor slot a problem? It's tempting to initialize all the unused descriptor slots to point to a purple error texture. Something has to check for out of range indices in shaders, but if unused slots are harmless, there's no need to check for a valid descriptor.

This can work. Thanks.

**Comment by cwfitzgerald** [comment]
> "The CPU timeline" concept needs to be clarified for multi-thread programs. There are potential locking bottlenecks. As noted, this is weird for multi-thread programs.

There's still a single timeline that all threads experience - that is, you will either see _before_ or _after_ the update. This is the same for all multi-threaded methods in wgpu.

> Right. That's why I was thinking in terms of an update queue. Number of changes per frame is probably < 100. Number of descriptors is on the order of 100,000 for a complex scene.

It's hard to avoid shadow copying _something_ - this could be optimized to as little as 4 bytes per descriptor though depending on implementation strategy. This will need to be driven by profiling.

> What prevents freeing a bound buffer? Something has to interlock against that. Will that be an error or an operation that is deferred until it is safe?

The same infrastructure that prevents it today. The bind group owns the texture (as in it actually just has an Arc<Texture>) so as long as we keep the command encoder alive until the gpu is done using it, the texture will stay alive. This is how it happens today.

> Is a shader accessing an unused descriptor slot a problem?

It will return unspecified (NOT undefined) results, but is valid.

> Something has to check for out of range indices in shaders

Yes, we will check against a metadata buffer, see the implementation notes in the spec

**Comment by John-Nagle** [comment]
> It's hard to avoid shadow copying _something_ - this could be optimized to as little as 4 bytes per descriptor though depending on implementation strategy. This will need to be driven by profiling.

Right. All binding updates commit at that copy, which allows for more optimization vs. bind-per-draw. Sounds good.

> (Referencing an unused descriptor slot) will return unspecified (NOT undefined) results, but is valid.

I'd suggest mapping unused texture descriptor slots to some built-in error texture, such as the purple often used for areas where nothing was drawn. Then errors become obvious. Mapping to a null handle means nothing is drawn, which is harder to debug.

**Comment by John-Nagle** [comment]
It looks like the design is ready. How far away is implementation and release? Thanks.

**Comment by cwfitzgerald** [comment]
Like all of wgpu contribution, it all depends on how much contributing effort is around, we can't promise any timelines.

**Comment by John-Nagle** [comment]
It looks like most of the support for bindless is in. Is there an example available yet? Preferably one where one thread adds and removes textures while the other renders. That would clarify how to use all this stuff properly. Thanks.

**Comment by kanerogers** [comment]
> Preferably one where one thread adds and removes textures while the other renders. 

Just out of curiosity, what's the link between multi-threading and bindless? I'm not sure I see an obvious connection.

**Comment by kanerogers** [comment]
argh, apologies, I had not intended to close this one @cwfitzgerald; misclicked! I seemingly do not have the permissions to re-open it.

**Comment by John-Nagle** [comment]
> Just out of curiosity, what's the link between multi-threading and bindless? I'm not sure I see an obvious connection.

Ah. The goal is to load content from one thread and command queue while concurrently rendering with another thread and command queue. Content loading should not impact rendering speed. (In single-thread mode, there's a huge impact on frame rate, like 5x.) It's possible to update the bindless descriptor table while the GPU is using it, provided that no non-null entry is touched during content loading. This should be the fastest way to update the GPU state. Vulkan supports this, and wgpu is getting close. I think. I'd like to see a simple example which shows proper fencing and locking for this. This is a difficult area and hard to get right.

**Comment by kanerogers** [comment]
Thanks for the detailed response. 

To be frank, that seems like a very strange setup. For context, I primarily work with Vulkan in mobile VR (a famously performance intensive environment) and in none of my titles have ever needed to implement a multi-threaded renderer.

Indeed, the point of bindless and the family of related techniques is to be more "GPU driven" and simply let the CPU get out of the way.

While I can see some potential CPU overhead in pipeline switching, rebinding of descriptor sets etc. (this is after all one of the motivators of a bindless architecture), it seems very odd that simply writing to a descriptor set has any kind of performance penalty for any real world game or content. I'd be very curious if you have profiler results indicating this to be the case. 

I say all this just to (a) understand your use case and (b) to try and set your expectations accordingly. Writing multi-threaded renderers is a very tricky business; indeed one that can be avoided by good architecture decisions.

**Comment by John-Nagle** [comment]
Re: kanerogers comment

A bit off-topic, but it's useful to see the motivation behind this.

There is a scaling issue that comes up for big open-world games. [Here's a video of what I'm doing.
](https://video.hardlimit.com/w/7usCE3v2RrWK6nuoSr4NHJ) That's a small part of one special event. The player is moving through a large, high-detail 3D world, with no constraints on their movement. That's all in Rust, using wgpu. 

If your scene, like a VR Beat Saber clone, is small enough that everything fits in GPU memory, you don't need concurrent asset loading. If you're doing a big world where the player can move around freely, you do. As the player moves through the world, content is constantly being loaded and unloaded from the GPU. Currently, I can usually get 60 FPS when the player is stationary and content loading is idle, but may drop to 10 FPS during heavy content loading. Most of the content loading work is done in other threads, but the copying into the GPU and the binding is all on one thread. That kills frame rate. I have lots of Tracy profiling data on this, and I've posted some of it elsewhere.

Very few people have done a high-detail open world game in Rust. Veloren, which is voxel-based, Hydrofoil Generation, which uses "good old DX11",  and my Sharpview are almost the only high-detail big-world systems in Rust.  So I'm the one who hits the scaling issues first.

Bevy has enough internal concurrency to use this approach effectively, but it's currently bottlenecked at one command queue, no bindless, as I understand it. So once this all works at the WGPU level, Bevy can move in that direction, too.

**Comment by kanerogers** [comment]
> If your scene, like a VR Beat Saber clone, is small enough that everything fits in GPU memory, you don't need concurrent asset loading. If you're doing a big world where the player can move around freely, you do. As the player moves through the world, content is constantly being loaded and unloaded from the GPU. 

Respectfully, my professional work includes more than an open source example I created 3 years ago. As a professional graphics programmer, I am well aware of the issues surrounding VRAM residency and streaming worlds, even on mobile.

For example, I recently delivered a client project, in Rust, that had a shared, mutable world that was practically infinitely large. You can see the talk I gave on it [here](https://www.youtube.com/watch?v=EB-ARcAnZY4). All rendering and data loading was performed on the main thread, and we shipped on target hardware running at 160hz.

> Currently, I can usually get 60 FPS when the player is stationary and content loading is idle, but may drop to 10 FPS during heavy content loading. Most of the content loading work is done in other threads, but the copying into the GPU and the binding is all on one thread. That kills frame rate. I have lots of Tracy profiling data on this, and I've posted some of it elsewhere.

This is avoiding my specific question and is still baking in the assumption that "multithreaded = faster". You may very well be introducing bottlenecks in your asset loading code with needless synchronisation. My exact question was: _are you seeing writing to descriptor sets in your profiler_. 

If you can provide more specific details, then your expectations can be set correctly. Bindless is not a magic bullet, and will not correct for poor architectural decisions.

**Comment by John-Nagle** [comment]
Replied via email; getting too far off topic.

**Comment by kanerogers** [comment]
> Replied via email; getting too far off topic. 

Gently, you have repeatedly asked for this feature, believing that it will solve a specific performance problem that you're facing. Insofar as your feature request is on topic, so too is a line of conversation verifying the logic of that request.

I am asking a single, specific and directed question: do you see descriptor writes taking a statistically significant portion of your frame time? If not, your performance limitations may be related to your overly complex multi-threaded renderer design. 

I am motivated to resolve your performance issues, but until that question can be answered, you repeatedly asking for the same feature will not result in any progress being made.

**Comment by John-Nagle** [comment]
Thanks.

It's great to hear from someone who pushed through big-world performance issues in a Rust->Vulkan environment. Not many people have done that.

Are descriptor writes taking up a substantial portion of time? No. Binding and unbinding is, because the current system doesn't have one big descriptor table for all the textures.

I don't have a multi-threaded renderer. I have a multi-threaded _application_. All the WGPU-level rendering is done from one thread.

For background, [see this issue from last fall](https://github.com/gfx-rs/wgpu/issues/6434#issuecomment-2475606011), where I provided a benchmark test case and Tracy traces. The basic result there is that adding and removing objects from outside the rendering thread has substantial impact on frame rate.  The goal is to get that interference down to a level that allows maintaining 60 FPS. Others can use this test case to reproduce the problem. "render-bench" has several branches so that different versions of wgpu can be compared. 

The stack I use is Rend3->EGUI->WGPU->Vulkan.  I have my own fork of Rend, as "rend3-hp", but that's just to keep up with the changes in all the other crates and fix some race conditions. 

The inner draw loop in Rend3 is basically bind, draw, unbind, for each object drawn. It's not using the one big bindless descriptor table approach.(Or so I think from looking at the code.) I think the bind part has a lock conflict with allocating new textures and meshes, but I'm not sure. Tracy isn't the right tool for finding that, because I have to fork Wgpu and put in Tracy profiling scopes to see that. I tried that once, but if Tracy profiling is done at a level of detail that lets me see this effect, it messes up the timing. A sampling profiler might help.

I watched [Kane Rogers-Wong's video.](https://www.youtube.com/watch?v=EB-ARcAnZY4) My takeaways from that were, oversimplifying:

- All this works better if you know Vulkan really well before you try using it with Rust.
- Wrapping a safe interface (such as WGPU or Vulkano) around Vulkan creates performance problems.
- Best performance is achieved by doing all the drawing without waiting for anything.

This may or may not be a good summary. Comments?

So my thinking for doing something similar atop WGPU is to put all the texture descriptors into one giant descriptor table, and avoid any binding or waiting during the draw process. Just queue up big blocks of commands for the GPU and submit them.

Then the problem is how to load content into the GPU in parallel with that. Since all new content comes in via other threads, I'd like to those operations use a separate transfer queue for the copy into the GPU. This creates the problem that the descriptor table may need to be modified while rendering is using it. (Or fully recopied for each frame, perhaps.)

What's the motivation for all this? It's the metaverse problem - how do we build big general-purpose 3D worlds with high detail, good performance, and user generated content? Available metaverses are either low-rez (Meta's Horizon, Decentraland), divided into small isolated areas, or sluggish (Second Life/Open Simulator). Or voxel-based. The voxel people do better at performance, but, of course, everything looks boxy.

A metaverse client has most of the problems of both an MMO client and a web client. You have to do most of the things an MMO client does. You have to do them on un-curated, and to some extent untrusted, content. Dealing with content overload while not killing the frame rate is tough but necessary.

What I'm writing is a new Second Life/Open Simulator client, [Sharpview](https://www.animats.com/sharpview). It's 100% safe Rust. It's far enough along that test users can log in and explore the 3D world. The entire world is the size of Los Angeles. There are 3 petabytes of content on the asset servers. 

Nobody has been able to crack the performance problem in the existing C++ clients. Nobody has tried serious multi-threading or Vulkan on this before, though. So that's what I'm working on.

(Probably longer than I should have, since the metaverse boom came and went. But I have free time and am not giving up.)

**Comment by kanerogers** [comment]
### In response to your comments

> It's great to hear from someone who pushed through big-world performance issues in a Rust->Vulkan environment. Not many people have done that.

Thank you, though as I hope will be clear as we go on, Rust is, for the most part, irrelevant here.

To begin with, I think we have some terminology confusion that may be impeding things. I should clarify that when I refer to a "multi-threaded renderer", my definition of the word "renderer" means "anything that makes GPU API calls". So when you say:

> I don't have a multi-threaded renderer. I have a multi-threaded application. All the WGPU-level rendering is done from one thread.

And follow it up with (emphasis mine)

> adding and removing objects from **outside the rendering thread** has substantial impact on frame rate.

This is fitting my definition of "a multi-threaded renderer", because you are making GPU API calls from more than one thread. Rendering (eg. recording draw calls to a command buffer and submitting them) is just one thing a renderer does.

Moving on,

> The stack I use is Rend3->EGUI->WGPU->Vulkan. I have my own fork of Rend, as "rend3-hp", but that's just to keep up with the changes in all the other crates and fix some race conditions.

A couple things there:

- Rend3 is now in maintenance mode. It would probably make sense for you to move away from it entirely
- That's a complicated stack. This makes it very difficult to track down exactly where the performance issues are coming from

> I watched [Kane Rogers-Wong's video.](https://www.youtube.com/watch?v=EB-ARcAnZY4) My takeaways from that were, oversimplifying:

I'm not quite sure why you're using my full name instead of just the pronoun "you", but, sure, yes that's my video that I linked you earlier.

> All this works better if you know Vulkan really well before you try using it with Rust.

Knowledge of a programming language (Rust) is unrelated to knowledge of a particular domain (Vulkan, GPU programming in general), so yes, absolutely, you will have better success if you have knowledge of a domain before working on a project in that domain. For context, I've been using Rust professionally since 2015, and Vulkan since 2020, so it's not a case of me having prior graphics experience before using Rust. Again, the programming language really is irrelevant here.

> Wrapping a safe interface (such as WGPU or Vulkano) around Vulkan creates performance problems.

Not exactly. What I said was that Rust's memory safety model is designed for the CPU; it does not very readily translate to the GPU, and as a result many Rust developers *poorly* and *misguidedly* implement that safety in high level wrappers.

WGPU, and the WebGPU project in general, have specific goals, as I explicitly called out in my talk, that require a safe execution environment when executing untrusted code (ie. running in a browser). There are trade-offs that come with this decision, but again, they make sense for the particular problem they were designed to solve.

> Best performance is achieved by doing all the drawing without waiting for anything.

Yes, things run faster when they're not waiting unnecessarily. On a GPU, it's a bit more complex because they are, by their very nature, massively parallel devices. Knowing *when* the GPU is waiting and *why* is harder to answer than on the CPU.

The core problem is trying to avoid two failure modes:
- Correct, but slow: waiting for every operation to complete before proceeding to the next
- Fast, but wrong: not waiting for any operation to complete before proceeding to the next

You'll find many talks [such as this one](https://www.youtube.com/watch?v=5GDg4OxkSEc&t=326s), [this one](https://www.youtube.com/watch?v=JvAIdtAZnAw&t=1354s) and [this one](https://www.youtube.com/watch?v=GiKbGWI4M-Y&t=2059s) that explain concepts like synchronisation, barriers and fences which can be used to ensure that the GPU is only waiting when it needs to wait, and no more. *Deeply* understanding this is an absolute requirement to writing high performance graphics code.

To your specific comments:

> So my thinking for doing something similar atop WGPU is to put all the texture descriptors into one giant descriptor table, and avoid any binding or waiting during the draw process. Just queue up big blocks of commands for the GPU and submit them.

Yes, this is a good strategy, with one caveat. You *must* ensure that an image has completed uploading and transitioning to the correct image layout before sampling it from a shader. See above for how to do that.

> Then the problem is how to load content into the GPU in parallel with that. Since all new content comes in via other threads, I'd like to those operations use a separate transfer queue for the copy into the GPU. 

Your use of the word "parallel" is imprecise here. This is a common mistake.

There is no correlation between the act of *recording* commands in parallel (ie. from two threads) and, what you really care about, *execution* of commands in parallel. So long as your hardware is capable (which, unless you're using truly ancient hardware will almost certainly be the case - if not, then nothing can help you), and you have recorded the correct synchronisation commands along with your transfer and draw commands (again, see above), then everything will execute *as parallel as possible*.

> This creates the problem that the descriptor table may need to be modified while rendering is using it. (Or fully recopied for each frame, perhaps.)

The specific issue of the descriptor table being modified *while another command is in flight that may still be using it*, is resolved by using "bindless" extensions. In the case of Vulkan, we can look at the documentation of the `VK_DESCRIPTOR_BINDING_UPDATE_AFTER_BIND_BIT` introduced in Vulkan 1.2. 

> `VK_DESCRIPTOR_BINDING_UPDATE_AFTER_BIND_BIT` specifies that if descriptors in this binding are updated between when the descriptor set is bound in a command buffer and when that command buffer is submitted to a queue, then the submission will use the most recently set descriptors for this binding and the updates do not invalidate the command buffer.

source: https://registry.khronos.org/vulkan/specs/latest/man/html/VkDescriptorBindingFlagBits.html

I have excluded the section on multi-threaded updates as I think it needlessly confuses things.

> What's the motivation for all this? It's the metaverse problem - how do we build big general-purpose 3D worlds with high detail, good performance, and user generated content? Available metaverses are either low-rez (Meta's Horizon, Decentraland), divided into small isolated areas, or sluggish (Second Life/Open Simulator). Or voxel-based. The voxel people do better at performance, but, of course, everything looks boxy.
> ...
> Nobody has been able to crack the performance problem in the existing C++ clients. Nobody has tried serious multi-threading or Vulkan on this before, though. So that's what I'm working on.

Finally, many very clever people have worked on simulating large virtual worlds, including at planetary scale, for over a decade. My conjecture for why you don't see many of these systems in the wild is that they (a) fundamentally do not deliver much value to players (bigger isn't better) and (b) have not solved the very real commercial problem of building an art and content pipeline at practically infinite scale. Rendering is only a *very* small part of the overall problem.

### Closing comments
You are, fundamentally, trying to tackle a very difficult problem. I think that your struggles with this problem are completely understandable, given the territory in question. 

My suggestion, coming from experience in having solved these same problems, is to *start small*. Strip away the layers of indirection between your application and the GPU, and start from first principles:

- How many objects do you need to render in a given frame? (ie. the "resident set")
- What portion of those objects does this represent out of all the possible objects that could be rendered? (ie. the "total set")
- How quickly can the resident set be swapped out? (ie. how fast do your players move, can they teleport, etc.)
- What is your total frame budget (eg. what is your target frame rate, on your target hardware)?
- How much of your frame time is consumed by drawing?
- How much time does this leave for transfers?
- Can you fit a total resident set swap, inside your transfer time budget?

Once you've framed the problem in these terms, you can start with ground truth (ie. how fast is your PCIe bus) and first determine whether what you're trying to achieve is even *theoretically possible*. From there, you can move on to specific techniques like compression, queued transfers, LODs, etc. etc. Perusal of the GDC vault should yield many successful solutions employed in large, open world games.

I hope the above was useful to you and helps you make some progress towards your desired goals.

**Comment by magcius** [comment]
Note that I believe the new bindless approach and strategy that we're aiming for in wgpu here *can* be both safe and relatively efficient, however it requires some care in how to use it effectively. Rend3 is, as of today, is simply not that renderer. This isn't a slight towards Connor or anyone else who worked on it; I think we'd both agree that Rend3's architecture is not where it needs to be to handle the scale you're throwing at it.

As I've said to you many times before, designing large-scale robust renderers is a complex problem in *any* technology stack, Rust or not, WGPU or not, Vulkan or not, and requires you to understand how to build and architect a renderer. It might be time to start designing a new one from scratch rather than hoping Rend3 magically gets there with WGPU upgrades.

**Comment by John-Nagle** [comment]
I'm painfully aware of the renderer problem. Various people have written My First Renderer about four times, but no pro-grade renderer has emerged. Here's what I know right now:

- Rend3 is _almost_ good enough for what I want to do. The API is well thought out. It's close to the three.js API. It's just tied to the way Vulkan and WGPU worked four or five years ago. It never got enough users to push it over the top. I have [my own fork of Rend3](https://github.com/John-Nagle/rend3-hp), and spend some time on speeding that up and fixing bugs. That's why I'm here in the WPGU forum.
- [Renderling](https://renderling.xyz/) has promise. That's on top of WGPU. It's one guy, but he has some EU funding.
- [Orbit](https://github.com/Thefefe/orbit) had promise, but seems to be abandoned. 
- Bevy's internal renderer I'm not sure about, but it is apparently hard to extract from its parent program without organizing everything around Bevy's ECS system.

> It might be time to start designing a new one from scratch.

You're welcome to try. A full rewrite leads to [second system syndrome](https://en.wikipedia.org/wiki/Second-system_effect). What's available isn't that bad. I'm a user of renderers, not an renderer developer.

Rend3 mostly works by having its API calls queue up instructions to change the scene. Before the next frame is drawn, those instructions are executed by the rendering thread, resulting in commands to the GPU. Then the CPU side of drawing takes place. (There's also a depth sort for translucency and a shadow pass in there.)

The "create mesh" and "create 2d texture" API calls, the big data move operations, do some work in the calling thread involving buffer allocation and copying. This seems to result in a lock conflict with the rendering thread, and rendering stalls. I think (but am not sure) that binding and buffer allocation clash somewhere. Anybody know? Incidentally, it doesn't matter if the big data move operations fall behind by a frame or two. They're not that time-critical. If drawing locks them out, no big deal.

So, using bindless mode means all the binding is done before rendering starts. Drawing can then (I think) run without acquiring locks during the draw phase. If I can get that to work, or someone else is interested in doing it, that would be enough for now.

I just need to get performance while asset loading is in progress up to where it is when asset loading is not in progress. There's about a 5x hit on performance, worst case, from that conflict now. This can be demonstrated with my [render-bench test.](https://github.com/John-Nagle/render-bench) That's a self-contained dummy scene which creates and destroys objects on a 10 second cycle, with long steady-state pauses. 100% CPU utilization on the render thread, maybe 30% GPU utilization.

**Comment by kanerogers** [comment]
> I'm painfully aware of the renderer problem. 

Respectfully, your comments to this point would indicate otherwise.

> You're welcome to try. A full rewrite leads to [second system syndrome](https://en.wikipedia.org/wiki/Second-system_effect). What's available isn't that bad. 

You are mistaken. There is no such thing as *a renderer*, in the general purpose sense (at least in real-time, anyway). No such software exists. Unity and Unreal are highly specialised pieces of software that, among many other things, perform rendering of their very specific data model. 

> I'm a user of renderers, not an renderer developer.

And here you are *doubly* mistaken. If you are trying to do what you're setting out to do (that is, render very large worlds at interactive framerates), you have, whether you like it or not, accepted the task of becoming a graphics developer. Personally, I find it quite enjoyable, I hope you do too.

You will have a far better time if you follow the advice I have laid out for you, than trying to find a "magic bullet" on the shelf. It does not exist.

> The "create mesh" and "create 2d texture" API calls, the big data move operations, do some work in the calling thread involving buffer allocation and copying. This seems to result in a lock conflict with the rendering thread, and rendering stalls. I think (but am not sure) that binding and buffer allocation clash somewhere. Anybody know? Incidentally, it doesn't matter if the big data move operations fall behind by a frame or two. They're not that time-critical. If drawing locks them out, no big deal.

Once again, you've veered away from discussing Vulkan, wgpu or even rendering, and are now talking about complex software that exists outside this repository. Please see my comments above.

> So, using bindless mode means all the binding is done before rendering starts. Drawing can then (I think) run without acquiring locks during the draw phase. If I can get that to work, or someone else is interested in doing it, that would be enough for now.

No. I explained this in a lot of detail above. The real problem is synchronisation and parallelism; bindless is but one part of the solution.

> I just need to get performance while asset loading is in progress up to where it is when asset loading is not in progress. There's about a 5x hit on performance, worst case, from that conflict now. This can be demonstrated with my [render-bench test.](https://github.com/John-Nagle/render-bench) That's a self-contained dummy scene which creates and destroys objects on a 10 second cycle, with long steady-state pauses. 100% CPU utilization on the render thread, maybe 30% GPU utilization.

It should go without saying, but this is again not relevant to wgpu, bindless or the complex task of GPU parallelism.

**Comment by magcius** [comment]
> I'm a user of renderers, not an renderer developer.

The problems you're running into are renderer problems, not user of renderer problems. This means you're going to need to do some amount of architecture work on your renderer in order to solve your problem, or pay someone else to solve your problems for you.

> So, using bindless mode means all the binding is done before rendering starts. Drawing can then (I think) run without acquiring locks during the draw phase. If I can get that to work, or someone else is interested in doing it, that would be enough for now.

This is not what bindless means, and even if all items in this Bindless Tracking PR were done, they would not solve your problems here.

**Comment by John-Nagle** [comment]
> There is no such thing as a renderer, in the general purpose sense (at least in real-time, anyway).

Hm? I just listed four of them.

The APIs of these mesh renderers have quite a bit in common. They all let you load in meshes, textures, skeletons, lights, materials, and transforms. Then you create an object which ties a set of those together, and something appears on screen. That's what Three.js, Rend3, and Renderling all offer. When you deserialize a glTF file, those are the data items you get. So things that can display glTF tend to be structured at some level with an API that accepts something close to those glTF data items.

The similarity of those APIs is important. It indicates things in that area are settling down towards a consensus.

(Voxel-based systems need a different model. As does Gaussian splatting.)

> It should go without saying, but this is again not relevant to wgpu, bindless or the complex task of GPU parallelism.

WGPU is a crate which offers a general API intended to be used by various kinds of renderers. So what renderers need from WGPU to do that job efficiently is relevant. WGPU is more than a safe wrapper around Vulkan. More tomorrow.

**Comment by kanerogers** [comment]
> Hm? I just listed four of them.

No, you didn't. You listed 4 projects that weren't a match for your requirements. Each have their own reasons for not meeting your requirements, *because no project will meet your requirements*.

The reason why many of these projects get abandoned is because the author, through no fault of their own, eventually realised that their initial goal was misguided. When they realised this, they stopped. Bevy falls into the "renderer for a game engine" case, which I explained earlier in the case of Unreal or Unity.

**There is no such thing as a general purpose renderer**.

This is *not* a controversial claim. I would encourage you to review a random selection of say, the top 1,000 games on Steam by sales. They will fall into one of three categories:

- AAA games that use proprietary, in-house renderers that have evolved over many years in conjunction with their in-house game engines.
- AAA to small indie games that use Unreal or Unity
- Indie games that use a single purpose, in-house game engine with its own renderer.

This data should, hopefully, in conjunction with the information you collected about some rendering projects, loudly indicate to you something very important: **there is no such thing as a renderer independent of a game engine**.
 
> The APIs of these mesh renderers have quite a bit in common. They all let you load in meshes, textures, skeletons, lights, materials, and transforms. Then you create an object which ties a set of those together, and something appears on screen. That's what Three.js, Rend3, and Renderling all offer. When you deserialize a glTF file, those are the data items you get. So things that can display glTF tend to be structured at some level with an API that accepts something close to those glTF data items.
> 
> The similarity of those APIs is important. It indicates things in that area are settling down towards a consensus.
> 
> (Voxel-based systems need a different model. As does Gaussian splatting.)

You are mistaking superficial similarity for general purpose re-usability. Please allow me to re-state, without reservation:

**There is no such thing as a general purpose renderer**.

The only way to refute this contention is with existence proof. I'd be more than happy to come back to you hat in hand should you be able to provide an example of such software used in production, and at scale.

> WGPU is a crate which offers a general API intended to be used by various kinds of renderers. So what renderers need from WGPU to do that job efficiently is relevant. WGPU is more than a safe wrapper around Vulkan. More tomorrow.

Correct. This would, by your own logic, mean that the best people to provide commentary on what features should be used by wgpu are the people creating their own renderers. Your continued misunderstanding of fundamental graphics concepts (because you are, by your own admission, not interested in doing this work yourself) leaves you ill-equipped to make specific feature recommendations.

This leaves you with a simple, binary choice, as clearly outlined to you by @magcius:

- Write your own renderer, from scratch, using wgpu
- Find a third-party to build the renderer you want

I am more than happy to discuss with you, at length, how to write a renderer that will solve the problems you are facing (though my expertise lies in Vulkan, not wgpu). However, unless you are willing to roll your sleeves up and do the work yourself, asking for specific features in a domain you don't understand is, and will remain, counterproductive to your stated goals.

Graphics programming may be intimidating at first, but hopefully as resources such as my talk would indicate, it is not an impossible challenge. If you have a large amount of free time on your hands, this would be an *excellent* way to spend it.

**Comment by John-Nagle** [comment]
Hm. [Wikipedia has a list of "higher level renderers".](https://en.wikipedia.org/wiki/List_of_3D_graphics_libraries). It's not that hard to do this if you don't need to scale up to large scenes. The big problems seem to appear when the renderer needs spatial data to decide what's visible and which lights can illuminate what. Rend3 tried occlusion culling, for example but it made performance worse in outdoor scenes. Renderling seems to have hit that problem too.

I agree, though, that nobody has really nailed how to decompose this problem well.

On a more useful note, back to what to do about bindless. 

What I need to speed up is [this inner loop in Rend3](https://github.com/John-Nagle/rend3-hp/blob/6e686ba4ffcdbf5bc1ecddbf0f325173ed1614ee/rend3-routine/src/forward.rs#L218C1-L240C14), which calls **set_bind_group** on each iteration. Something in there can be stalled by texture creation or buffer allocation taking place in another thread. 

I've tried adding more detailed profiling scopes, but doing that in this inner loop affects performance enough to confuse the issue.

My interest in bindless is basically to speed up this inner loop, by moving anything that can block out of the inner loop.

Is that reasonable?

**Comment by magcius** [comment]
I'm glad you've found a specific bottleneck to take a closer look at. It sounds like that inner loop "slows down" when you're loading assets on a different thread. There are some hypotheses worth investigating:

1. The bottleneck is in Rend3's architecture; loading a mesh/material/whatever causes the inner loop to slow down.
2. The bottleneck is in wgpu's architecture; creating/uploading wgpu's buffers and textures causes set_bind_group to slow down.
3. The bottleneck is in your driver; creating/uploading D3D12/Vulkan's buffers and textures causes things to slow down.

One test to evaluate which scenario you're in is to be in a stable state where you're in a world not doing any streaming, and then try activating creating/uploading raw wgpu buffers or textures, and see if that has the same slowdown. You can then also look into creating bind groups and see if you encounter the same slowdown. Using this approach, you can try to narrow down where the bottlenecks are.

The most likely scenario is likely 1., that there's a bottleneck in Rend3's architecture, perhaps some internal locks in the material system that you're running into with contention. But I wouldn't put it past it that some internal implementation detail of wgpu is where the issue lies. If the tests show you've identified a bottleneck in wgpu in particular, I'm sure the maintainers would love to hear about it, especially if it comes with a small reproducible example.

Rend3 does have a path where that set_bind_group is outside of the loop, that's its GPU-driven rendering feature, and it uses wgpu's existing bindless paths. That has some scalability issues that this issue's tasks are trying to fix; the existing system creates one large bind group, but creating this bind group can be somewhat expensive as materials and textures stream in and out.

**Comment by John-Nagle** [comment]
Just as a note on progress in profiling:

I'm currently looking at output from **samply** profiling. (Tracy doesn't go deep enough without adding more profiling scopes that slow things down. `cargo flamegraph` doesn't find any wide flames indicating hot spots. Samply samples and gives a nested list, and can be focused on parts of the timeline that are of interest.)

First results, not to be taken as definitive.

30% of time in my render-bench program is going into translucency depth sorting. I knew that. That's mostly a property of render-bench, where everything is marked "blend", the worst case. Also, building up the sort keys in Rend3 is more expensive than it should be, sometimes taking up more time than the sort.

24% of time in my render-bench program is going into an access function, `material()` here:

https://github.com/John-Nagle/rend3-hp/blob/6e686ba4ffcdbf5bc1ecddbf0f325173ed1614ee/rend3/src/managers/material.rs#L76C1-L84C2

That ought to be trivial. Puzzled as to how that can use so much time. The array being indexed is a list_any::VecAny, which involves downcasting. Need to dig further. More later.

WGPU's draw_indexed, set_bind_group, create_buffer, and builder::build add up to about 7%. That's fine; they're doing real work. Probably not encountering a lock or fence delay at the WGPU level. That's good to know.

**Comment by jimblandy** [comment]
@John-Nagle @kanerogers @magcius This issue is not the appropriate place for this discussion. We need this issue for tracking bindless design and implementation work, not debating application architecture and sharing profiling advice. Please file separate issues or open a Discussion if you'd like to continue the conversation.

**Comment by jimblandy** [comment]
> Replied via email; getting too far off topic.

@John-Nagle Thanks for redirecting the conversation to a more appropriate venue - it is much appreciated.

**Comment by IsaacMarovitz** [comment]
I'm currently working on an application that needs bindless, and I primarily work on Mac, so I may look at tackling the required integrations for Metal. It's fairly simple to do using argument buffers.
